common/dev/xdpe12284c.c

```diff
@@ -158,6 +193,391 @@ static float vid_to_float(int val, uint8_t vout_mode)
 	return 0;
 }
 
+static bool parsing_image(uint8_t *hex_buff, struct xdpe_config *dev_cfg)
+{
+	CHECK_NULL_ARG_WITH_RETURN(hex_buff, false);
+	CHECK_NULL_ARG_WITH_RETURN(dev_cfg, false);
+
+	bool ret = false;
+	int max_data_size = MAX_CMD_LINE;
+	dev_cfg->data = (uint8_t *)malloc(max_data_size);
+	if (!dev_cfg->data) {
+		LOG_ERR("Failed to malloc data!");
+		goto exit;
+	}
+
+	dev_cfg->crc_exp = 0;
+
+	bool rec_flag = false;
+	bool new_line = true;
+	bool got_end_flag = false;
+	uint16_t ofst = 0;
+	uint16_t value = 0;
+	int data_idx = 0, val;
+	for (int i = 0; i < fw_update_cfg.image_size; i++) {
+		/* check valid */
+		if (!hex_buff[i]) {
+			LOG_ERR("Get invalid buffer data at index %d", i);
+			goto exit;
+		}
+
+		/* collect data */
+		if (rec_flag == true) {
+			// grep offset and exit keyword
+			if (new_line == true) {
+				val = ascii_to_val(hex_buff[i]);
+				if (val == -1) {
+					LOG_ERR("Image got format error in line %d", __LINE__);
+					goto exit;
+				}
+				if (val != 2) {
+					if (i + 3 < fw_update_cfg.image_size) {
+						if (!strncmp(&hex_buff[i], "[End", 4)) {
+							got_end_flag = true;
+							break;
+						}
+					}
+					/* Assume there's no other key to access */
+					got_end_flag = true;
+					break;
+				} else {
+					if (i + 3 < fw_update_cfg.image_size) {
+						ofst = 0;
+						for (int j = i; j < (i + 4); j++) {
+							val = ascii_to_val(hex_buff[j]);
+							if (val == -1) {
+								LOG_ERR("Image got format error in line %d",
+									__LINE__);
+								goto exit;
+							}
+							ofst = (ofst << 4) | val;
+						}
+						i += 3; //pass next 4 bytes offset
+					}
+				}
+				new_line = false;
+			} else {
+				if (hex_buff[i] == ' ') {
+					if (i + 4 >= fw_update_cfg.image_size) {
+						LOG_ERR("Image got format error in line %d",
+							__LINE__);
+						goto exit;
+					}
+
+					// skip collect empty data '----'
+					if (hex_buff[i + 1] == '-') {
+						ofst++;
+						i += 4; //pass '----'
+						continue;
+					}
+
+					value = 0;
+					for (int j = i + 1; j < (i + 5); j++) {
+						val = ascii_to_val(hex_buff[j]);
+						if (val == -1) {
+							LOG_ERR("Image got format error in line %d",
+								__LINE__);
+							goto exit;
+						}
+						value = (value << 4) | val;
+					}
+
+					if (data_idx + 3 >= MAX_CMD_LINE) {
+						LOG_ERR("Data collect over limit size %d",
+							max_data_size);
+						goto exit;
+					}
+					LOG_DBG("collect new data ofst: 0x%x val: 0x%x", ofst,
+						value);
+					memcpy(&dev_cfg->data[data_idx], &ofst, 2);
+					memcpy(&dev_cfg->data[data_idx + 2], &value, 2);
+					data_idx += 4;
+					ofst++;
+					i += 4; //pass 4 bytes data
+				} else if (hex_buff[i] == 0x0d) {
+					// '\n'
+					i++;
+					new_line = true;
+				} else {
+					LOG_ERR("Image got format error in line %d", __LINE__);
+					goto exit;
+				}
+			}
+		}
+
+		/* parsing address and crc */
+		if (i + 9 < fw_update_cfg.image_size) {
```

Comment:
This needs to be refactored as 7 level deep if statements are not easily maintained.